### PR TITLE
Azure Function pipeline template

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR adds a pipeline template that can deploy an Azure Function. It has a few conventions baked into it and it basically extracts the template that I've created in this PR: #487 

I'll be using this template with Check Enforcer to get some rigor into its deployment process, and I can use it with any other automation that we choose to develop.

### CODE REVIEW

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+parameters:
+- name: ToolName
+  type: string
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:3]: Isn't string the default type? What value do we gain by explicitly listing the type? Not pushing back but just trying to understand when someone might want to do this.

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      - job: BuildPackage
+        pool:
+          vmImage: ubuntu-16.04
+        steps:
+          - task: UseDotNet@2
+            displayName: Tools
+            inputs:
+              version: '3.1.201'
+          - pwsh: |
+              dotnet restore tools/${{ parameters.ToolName }}
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:27]: Why separate this from build? build will do it automatically.

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            runOnce:
+              deploy:
+                steps:
+                - task: ArchiveFiles@2
+                  displayName: Archive
+                  inputs:
+                    rootFolderOrFile: $(Pipeline.Workspace)/artifacts/app
+                    includeRootFolder: false
+                    archiveFile: $(Pipeline.Workspace)/artifacts/${{ parameters.ToolName }}-$(Build.BuildID).zip
+                - task: AzureFunctionApp@1
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:61]: Shouldn't we also sign these?

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            condition: succeededOrFailed()
+
+    - ${{if and(ne(variables['Build.Reason'], 'PullRequest'), eq(variables['System.TeamProject'], 'internal'))}}:
+      - stage: Staging
+        dependsOn: Build
+        jobs:
+        - deployment: PublishFunction
+          environment: ${{ parameters.StagingEnvironmentName }}
+          pool:
+            vmImage: ubuntu-16.04
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:50]: Should this be ubuntu-18.04?

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml (hunk 1 above)
Comment (mitchdenny) [reply eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:3]: It is just me being explicit. I also prefix local member access in C# with ```this```. Don't hate me :) I'll switch it to being implcit.

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml (hunk 2 above)
Comment (mitchdenny) [reply eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:27]: Will collapse them together.

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml (hunk 4 above)
Comment (mitchdenny) [reply eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:50]: I'll give Ubuntu 18.04 a run.

File: eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml (hunk 3 above)
Comment (mitchdenny) [reply eng/pipelines/templates/stages/archetype-sdk-tool-azure-function.yml:61]: I'm looking into this. We can probably sign it but need to think through whether there is a benefit. We aren't publishing the artifacts anywhere, they are running in production and whilst it is nice for them to be signed I'm not sure Azure Functions can enforce code integrity based on code signing.

Looking into it with the team that owns some of that tooling.

**Pull request merged**